### ISSUE TITLE
Disable or Move MixinProfiler to a MixinModule/Plugin

### ISSUE THREAD

**Issue by ryantheleach** [SpongePowered/Sponge#1726]
At least one mod (TickDynamic) uses the Vanilla profiler to use measurements to change the behavior of the server.

By moving the MixinProfiler to something that can be enabled for performance (if never used) in the performance config settings, we can allow some compatibility with TickDynamic.

In addition @pokechu22 has requested that we not mess with the vanilla profiler, as it's one method of directly comparing SpongeVanilla's performance compared to MojangVanilla. (via https://minecraft.gamepedia.com/Commands/debug)

@gabizou Tagging you as you authored this Mixin and may have some idea about the perf difference.

**Comment by phit** [comment]
why isn't this already connected to the timings module config that exists in the `global.conf`?

**Comment by gabizou** [comment]
The profiler itself is a poor mans profiler in comparison to timings and traditional JVM profiling wit a proper profiler. Yes, we can enable it to allow vanilla comparison of performance, but there are several specific calls that degrade performance. In the context of resolving the TickDynamic mod and including the responsible code into our codebase, it is 100% absolutely not necessary for us to use the profiler itself to achieve the same results. We have near absolute control over the ticking pipeline, and can even additionally replace things like TickLrofiler if we wanted to associate the processing requirements that a particulR entity or tile entity is occupying. 

In short, I’m grossly opposed to making it a module and reinstating the commented out profiler calls in tick methods. However, I’m open for debate, if @Zidane or @bloodmc are wanting to provide devils advocate.

**Comment by Pokechu22** [comment]
I don't see any issue with making it configurable.  Yes, it's expensive to have it running most of the time, but keeping it disabled by default but allowing users to enable it has no downside other than the work.

Also, the issues with calls being expensive has actually been fixed in 1.12.1 (ref [MC-117087](https://bugs.mojang.com/browse/MC-117087 "[MC-117087] Class.getSimpleName performance issue with large number of block entities")).  There are now methods where you can pass `Supplier<String>` to the profiler and it'll only evaluate it if the profiler is enabled (which, notably, you have not stubbed out, although they only call methods that you have stubbed).  These methods are not currently named (because MCPBot only is doing 1.12.0 mappings :/) but are `func_194340_a` (mirrors `startSection`) and `func_194339_b` (mirrors `endStartSection`).  But I think they address the performance concerns fairly well.

**Comment by Pokechu22** [comment]
I've encountered a second vanilla use case: in 1.13, commands are now in the profiler, and I've already seen cases where users are using that (mainly for functions), for instance [MC-124838](https://bugs.mojang.com/browse/MC-124838).  While the profiler isn't great, some users are using it for great benefit, and I don't _think_ the timings profiler will fill that use case.

**Comment by Pokechu22** [comment]
Another point: I just learned sponge can be used in single-player; in that case breaking the profiler breaks the vanilla shift+f3 pie chart profiler.

**Comment by Pokechu22** [comment]
Just tested it.  The game crashes when Shift+F3 is used to open the profiler, because it's always returning an empty list and the game assumes that if profiling is enabled, it won't be empty.

[Crash report](https://gist.github.com/Pokechu22/99b69a1e40f15e12f61e6f4a045b593b)

**Comment by LemADEC** [comment]
Could we at least disable it while it's crashing until we get a proper fix, please?

**Comment by Pokechu22** [comment]
It (the profiler) being disabled is what is causing the crash.  (On rereading I see you could be referring to the mixin too; disabling that would probably fix the crash but I'm not 100% sure)

**Comment by LemADEC** [comment]
Let me reformulate: crashing client is bad, we should prevent it at least with a temporary fix while a long term solution is being worked on.

**Comment by gabizou** [comment]
I'm re-enabling the profiler in #1951 since I had to re-sync and verify a bunch of vanilla code in Sponge's code.

**Comment by ImMorpheus** [comment]
So... Is the vanilla profiler enabled now ? Can this be closed ?

**Comment by phit** [comment]
![image](https://user-images.githubusercontent.com/2097483/42183922-e0f0b176-7e43-11e8-91b1-41fec6a3c2b8.png)
